### PR TITLE
trace: Add program-wide default dispatcher

### PR DESCRIPTION
**Pull request by kazimuth**
## Motivation

I was just trying to use tokio-trace for a greenfield project, but I was frustrated to discover that I couldn't really use it easily.

I was using the [`runtime`](https://docs.rs/runtime/0.3.0-alpha.4/runtime/) crate, which transparently spawns a thread pool executor for futures. In that thread pool, there's no way to set a tokio-trace subscriber for the duration of each thread, since you don't control the thread initialization. You *might* be able to wrap every future you spawn with a subscriber call, but that's a lot of work.

I was also confused because the documentation said that setting a subscriber in the main thread would use that subscriber for the rest of the program. That isn't the case, though -- the subscriber will be used only on the main thread, and not on worker threads, etc.

## Solution

I added a function `set_global_default`, which works similarly to the `log` crate:

```rust
tokio_trace::subscriber::set_global_default(FooSubscriber::new());
```

The global subscriber (actually a global `Dispatch`) is a `static mut` protected by an atomic; implementation is copied from the `log` crate. It is used as a fallback if a thread has no `Dispatch` currently set. This is extremely simple to use, and doesn't break any existing functionality.

Performance-wise, thread-local `Dispatch` lookup goes from ~4.5ns to ~5ns, according to the benchmarks. So, barely any runtime overhead. (Presumably there's a little compile-time overhead but idk how to measure that.) Since the atomic guard is only ever written once, it will be shared among a CPU's cores and read very cheaply.

I added some docs to partially address #1151. I also switched the tokio-trace benchmarks to criterion because the nightly benchmarks weren't compiling (missing `dyn` flags?)

### CODE REVIEW

File: tokio-trace/Cargo.toml [hunk 1]
```diff
@@ -21,7 +22,7 @@ categories = ["development-tools::debugging", "asynchronous"]
 keywords = ["logging", "tracing"]
 
 [dependencies]
-tokio-trace-core = "0.2"
+tokio-trace-core = { version = "0.2", path = "tokio-trace-core" }
```
Comment (kazimuth) [inline tokio-trace/Cargo.toml]: this change was for testing, i can take it out if there's some other way to set this up

File: tokio-trace/src/subscriber.rs [hunk 2]
```diff
@@ -17,3 +17,19 @@ where
...
+/// Sets this subscriber as the global default for the duration of the entire program.
+/// Will be used as a fallback if no thread-local subscriber has been set in a thread (using `with_default`.)
+///
+/// Can only be set once; subsequent attempts to set the global default will fail.
+/// Returns whether the initialization was successful.
+///
+/// [span]: ../span/index.html
+/// [`Subscriber`]: ../subscriber/trait.Subscriber.html
+/// [`Event`]: ../event/struct.Event.html
+pub fn set_global_default<S>(subscriber: S) -> bool
```
Comment (davidbarsky) [inline tokio-trace/src/subscriber.rs]: I think I'd prefer if this function returned a `Result<(), InitializationError>`, as `Result` is `#[must_use]`, which introduces higher misuse resistance (e.g., a users forgets to check if `set_global_default()` returns false). With `Result`, one could write `set_global_default(SomeSubscriber).expect("Unable to set default subscriber");` and crash immediately if there is some sort of logic failure (e.g., a global subscriber was set twice).

**Comment by hawkw** [comment]
Thanks! There's recently been some discussion in the [#dev channel](https://gitter.im/tokio-rs/dev) on Gitter about adding a feature like this, so it's great to see a PR. I think @jonhoo will be pleased to see this.

I'll probably have a chance to review this more thoroughly later this weekend or on Monday.

File: tokio-trace/tokio-trace-core/src/dispatcher.rs [hunk 3]
```diff
@@ -484,4 +526,65 @@ mod test {
...
+                span::Id::from_u64(1)
+            }
+            fn record(&self, _: &span::Id, _: &span::Record) {}
+            fn record_follows_from(&self, _: &span::Id, _: &span::Id) {}
+            fn event(&self, _: &Event) {}
+            fn enter(&self, _: &span::Id) {}
+            fn exit(&self, _: &span::Id) {}
+        }
+
+        // NOTE: can't have multiple tests that set the global dispatcher
```
Comment (hawkw) [inline tokio-trace/tokio-trace-core/src/dispatcher.rs]: I think that if these tests were in the integration tests (`tokio-trace-core/tests`), rather than in `src`, you could have multiple tests, if each test was its own separate file in `tests`. This is because each file in `tests` is compiled into its own binary when run, and will thus have its own globals.

We use a similar approach for testing filter caching, since that involves the global registry. You can see that in use [here](https://github.com/tokio-rs/tokio/tree/master/tokio-trace/tests).

File: tokio-trace/Cargo.toml [hunk 4]
```diff
@@ -30,6 +31,7 @@ ansi_term = "0.11"
 humantime = "1.1.1"
 futures = "0.1"
 log = "0.4"
+criterion = { version = "0.2", default_features = false }
```
Comment (hawkw) [inline tokio-trace/Cargo.toml]: I'm not _opposed_ to using `criterion` for `tokio-trace`'s benchmarks, but I'd like the motivation to be recorded in the PR description. Also, I think a change to switch to `criterion` should be moved to a separate PR, since we tend to merge PRs by squashing them --- since it's not directly related to the rest of the changes in this PR, it should have its own own commit in the master git history.

File: tokio-trace/README.md [hunk 5]
```diff
@@ -128,9 +128,25 @@ In order to record trace events, executables have to use a `Subscriber`
 implementation compatible with `tokio-trace`. A `Subscriber` implements a way of
 collecting trace data, such as by logging it to standard output.
 
-Unlike the `log` crate, `tokio-trace` does *not* use a global `Subscriber` which
-is initialized once. Instead, it follows the `tokio` pattern of executing code
-in a context. For example:
+There currently aren't too many subscribers to choose from. The best one to use right now
```
Comment (hawkw) [inline tokio-trace/README.md]: The `lib.rs` top-level docs currently include most of the same content as `README.md`; it would be good to make this change there as well.

File: tokio-trace/README.md [hunk 6]
```diff
@@ -14,10 +14,10 @@ often be quite challenging. Since individual tasks are multiplexed on the same
 thread, associated events and log lines are intermixed making it difficult to
 trace the logic flow. `tokio-trace` expands upon logging-style diagnostics by
 allowing libraries and applications to record structured events with additional
-information about *temporality* and *causality* — unlike a log message, a span
+information about _temporality_ and _causality_ — unlike a log message, a span
```
Comment (hawkw) [inline tokio-trace/README.md]: This looks like an unrelated change?

File: tokio-trace/tokio-trace-core/src/dispatcher.rs [hunk 7]
```diff
@@ -63,6 +73,30 @@ pub fn with_default<T>(dispatcher: &Dispatch, f: impl FnOnce() -> T) -> T {
     let _guard = State::set_default(dispatcher.clone());
     f()
 }
+
+/// Sets this dispatch as the global default for the duration of the entire program.
+/// Will be used as a fallback if no thread-local dispatch has been set in a thread
+/// (using `with_default`.)
+///
+/// Can only be set once; subsequent attempts to set the global default will fail.
+/// Returns Err if the global default has already been set.
```
Comment (hawkw) [inline tokio-trace/tokio-trace-core/src/dispatcher.rs]: I think it's worth cautioning that libraries should generally not call this function, to avoid conflicting with a global subscriber set by their dependents.

File: tokio-trace/tokio-trace-core/src/dispatcher.rs [hunk 8]
```diff
@@ -89,7 +123,15 @@ where
         .try_with(|state| {
             if state.can_enter.replace(false) {
                 let _guard = Entered(&state.can_enter);
-                f(&state.default.borrow())
+
+                let default = &state.default.borrow();
+
+                if default.is::<NoSubscriber>() && GLOBAL_INIT.load(Ordering::SeqCst) == INITIALIZED
```
Comment (hawkw) [inline tokio-trace/tokio-trace-core/src/dispatcher.rs]: Should we consider setting the thread-local default to the global default in this case? That might improve performance on subsequent calls to `get_default`. Of course, we would want to test this.

File: tokio-trace/tokio-trace-core/src/dispatcher.rs [hunk 9]
```diff
@@ -63,6 +73,30 @@ pub fn with_default<T>(dispatcher: &Dispatch, f: impl FnOnce() -> T) -> T {
...
+/// Will be used as a fallback if no thread-local dispatch has been set in a thread
+/// (using `with_default`.)
+///
+/// Can only be set once; subsequent attempts to set the global default will fail.
+/// Returns Err if the global default has already been set.
+///
+/// [span]: ../span/index.html
+/// [`Subscriber`]: ../subscriber/trait.Subscriber.html
+/// [`Event`]: ../event/struct.Event.html
+pub fn set_global_default(dispatcher: Dispatch) -> Result<(), ()> {
```
Comment (hawkw) [inline tokio-trace/tokio-trace-core/src/dispatcher.rs]: This should probably return a type implementing `std::error::Error` --- something like 
```rust
pub struct SetDefaultError {
    _p: () // a private field prevents user code from constructing this type
}

impl fmt::Display for SetDefaultError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.pad("a global default trace dispatcher has already been set")
    }
}

impl error::Error for SetDefaultError {}
```

File: tokio-trace/README.md (hunk 6 above)
Comment (kazimuth) [reply tokio-trace/README.md]: ah, that's my editor's autoformat, whoops

File: tokio-trace/tokio-trace-core/src/dispatcher.rs (hunk 8 above)
Comment (kazimuth) [reply tokio-trace/tokio-trace-core/src/dispatcher.rs]: Oh, good point. Yes, that probably will improve performance, will do.

File: tokio-trace/Cargo.toml (hunk 4 above)
Comment (kazimuth) [reply tokio-trace/Cargo.toml]: I can revert the change; just FYI though, the benchmarks aren't compiling on current nightly due to missing `dyn` annotations. I can add those in, but I wasn't sure if it would screw with backwards-compatibility requirements.

File: tokio-trace/tokio-trace-core/src/dispatcher.rs (hunk 8 above)
Comment (kazimuth) [reply tokio-trace/tokio-trace-core/src/dispatcher.rs]: Correctness sketch: because the global default, once set, cannot be changed, all subsequent hits to thread_local NoSubscribers will always fall through to the global default. So, it's fine to cache it locally.

File: tokio-trace/Cargo.toml (hunk 4 above)
Comment (hawkw) [reply tokio-trace/Cargo.toml]: We should probably add an `allow` attribute to fix the benchmarks on the latest nightly, but they don't get built on the CI builds that run against earlier Rust versions (since those are stable), so it would probably also be fine to just add `dyn`. 

Again, I'm fine with switching to `criterion` (and being able to run the benches without needing a nightly compiler would make my life easier, too 😄) but we should probably do it in a separate branch. Sorry for being persnickety, I just like having a clean git history on master.

File: tokio-trace/tokio-trace-core/src/dispatcher.rs (hunk 8 above)
Comment (hawkw) [reply tokio-trace/tokio-trace-core/src/dispatcher.rs]: We could also check if there's a global default when _exiting_ a scoped default dispatcher, and set the thread-local to that if there is, rather than always setting it back to `NoSubscriber`. I'd expect entering/exiting a dispatcher scope to not be very common, but _getting_ the default is fairly hot.

File: tokio-trace/Cargo.toml (hunk 4 above)
Comment (hawkw) [reply tokio-trace/Cargo.toml]: Thanks for moving this change to #1163.

**Comment by kazimuth** [comment]
This should be good to merge pending review from @carllerche

**Comment by carllerche** [comment]
If @hawkw is :+1: then I am OK w/ this.

**Comment by hawkw** [comment]
@kazimuth just checking if you were planning on making the handful of style changes I suggested? If you're not, I'll go ahead and merge. Otherwise, if you want to make the changes but don't have time, I'm happy to make the changes myself before merging?

**Comment by kazimuth** [comment]
I was actually going to do them this afternoon :)

**Comment by hawkw** [comment]
Excellent, thanks! Just making sure I shouldn't merge the PR now.
